### ISSUE TITLE
[Feature Request] Ignore properties when binding with Options service

### ISSUE THREAD

**Issue by lonix1** [dotnet/runtime#58384]
### Is your feature request related to a problem? Please describe.

I'm using the Options service to bind hierarchical [configuration](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-5.0#bind-hierarchical-configuration) data.

Suppose I have this class:
````cs
public class PersonOptions
{

  public string Name { get; set; }

  public int Age { get; set; }

  public bool Foo {                                // <--- I want binder to ignore this
    get {
      // ... do stuff
    }
  }

  public bool IgnoreMe => throw new Exception();   // <--- I want binder to ignore this

}
````

Because the `Foo` property is `public`, the binder runs it (which leads to various problems - why is it run?). My design requires it to be `public`, and I prefer not to change it to a method.

I hoped I could force the binder to ignore that property, e.g. using an attribute. I tried `[JsonIgnore]` but that didn't work.

### Describe the solution you'd like
Please consider adding an attribute for this - if the property is decorated by that attribute, then don't try bind it, and don't run it.

### Additional context
Thank you!

**Comment by pfeigl** [comment]
Why has this been dropped from 7.0? 

Looking at the `BinderOptions` property
```csharp
        /// <summary>
        /// When false (the default), the binder will only attempt to set public properties.
        /// If true, the binder will attempt to set all non read-only properties.
        /// </summary>
        public bool BindNonPublicProperties { get; set; }
```

This issue even looks more like a bug than a feature request, because a property with only a getter is `read-only` from the perspective of the binder.

The code behind the `ConfigurationBinder` changed quite some bit but I think based on these lines it'll still suffer from the same problemd.
```csharp

                // If we don't have an instance, try to create one
                if (bindingPoint.Value is null)
                {
                    // if the binding point doesn't let us set a new instance, there's nothing more we can do
                    if (bindingPoint.IsReadOnly)
                    {
                        return;
                    }
```` 

It'll first try to access the bindingpoint value (which trys to resolve via the property getter) before it checks on the `IsReadOnly`. By doing the check earlier, the whole problem would resolve itself I think, because it wouldn't even read the value, when it's not going to set it anyways.

**Comment by lonix1** [comment]
Agreed, it doesn't work as expected, it's a bug. And I still don't understand why it *runs* my non-bindable properties.

If you have a better name for the issue so it gets more attention, let me know and I'll rename it.

**Comment by turowicz** [comment]
They should use System.Text.Json to read the config files... would make it so much easier.

**Comment by ondrejtucny** [comment]
This is clearly a short-coming. A `ConfigurationIgnoreAttribute` should be added for sure; at least for the sake of parity considering `ConfigurationKeyNameAttribute` was implemented (see #36010).

Simple cases for supporting this cause:

- backwards compatibility
- post-load value conversion
- post-load options derivatives

Example — my current issue due to the absence of such an attribute is the need to bind values specified in a nested `IConfigurationSection` into a pre-created object followed by additional processing of the loaded values and selection of some specific defaults (code simplified; in the real case the post-processing is more complicated):

```
    /// <summary>
    /// CSV ingestion options describing various CSV formats ingested (loaded) by the system.
    /// </summary>
    public class CsvIngestionOptions
    {
        /// <summary>
        /// Default CSV format options.
        /// </summary>
        [ConfigurationKeyName("__ingored_"+nameof(DefaultFormat))] // BUG .NET has no attribute to ignore a particular property; see also https://github.com/dotnet/runtime/issues/58384
        public CsvFormatOptions DefaultFormat { get; set; }

        /// <summary>
        /// Default CSV format options; for the sake of sparse JSON deserialization.
        /// </summary>
        [ConfigurationKeyName(nameof(DefaultFormat))]
        public IConfigurationSection? DefaultFormatSection { get; set; }

        /// <summary>
        /// Invoked when the configuration is loaded from the file.
        /// </summary>
        internal void OnConfigurationLoaded()
        {
            // load default format, ensuring that at least the default encoding is set
            DefaultFormat = new CsvFormatOptions();
            if (DefaultFormatSection != null) DefaultFormatSection.Bind(DefaultFormat);
            DefaultFormat.EnsureEncodingDefined();
        }
    }
```

Note: For the options being loaded from a JSON file, I would assume things as `[JsonIgnore]`, `[JsonPropertyName]`, and `IJsonOnDeserialized` to work naturally. However, they don't. This behavior deserves a mention in the docs, at least.

**Comment by odyth** [comment]
The `[BindNever]` attribute also doesn't work to suppress the all powerful binder from evaluating readonly properties as of .netcore 6

**Comment by crozone** [comment]
Any movement on this?

**Comment by svick** [comment]
@lonix1 Yes, read-only properties are not read, but a "don't bind this" attribute doesn't exist. (We now have https://github.com/dotnet/runtime/issues/125111 open for that, but we haven't talked about that yet.)
